### PR TITLE
[Core][5/N] Fully working chunked prefill e2e

### PR DESCRIPTION
**Pull request by rkooo567**
This PR is a part of the RFC https://github.com/vllm-project/vllm/issues/3130.

This PR enables chunked prefill e2e. Note that chunked prefill is an experimental feature now (though it is actively used within Anyscale), and I will start serious benchmark with this PR. 

The feature can be enabled by using `enable_chunked_prefill`. The chunking is done based on `max_num_batched_tokens` (a.k.a token budget). This is the same way as described in SARATHI paper https://arxiv.org/abs/2308.16369. It also means we can have up to maximum 2 chunked prefill at any given time.

This PR
- Allow to put 2 attention metadata, one for prefill and one for decode.
    - it is done by broadcasting metadata twice. It can be more optimized by coelescing tensors better, but I made this way for simplicity. The normal path should just use 1 broadcast as usual. 
- Allow to run attention backend when there are prefill and decode mixed up.
- Ignore the generated token if chunked prefill is enabled.
- Add a chunked prefill test with various chunk size, cuda graph, and tp settings.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/config.py [hunk 1]
```diff
@@ -563,9 +563,13 @@ def __init__(
         if max_num_batched_tokens is not None:
             self.max_num_batched_tokens = max_num_batched_tokens
         else:
-            # If max_model_len is too short, use 2048 as the default value for
-            # higher throughput.
-            self.max_num_batched_tokens = max(max_model_len, 2048)
+            if enable_chunked_prefill:
+                # For chunked prefill, choose the well-tuned batch size.
```
Comment (rkooo567) [inline vllm/config.py:567]: This is arbitrary value. We should profile again.

File: vllm/core/scheduler.py [hunk 2]
```diff
@@ -140,7 +140,12 @@ def _sort_by_lora_ids(self) -> bool:
 
     @property
     def lora_requests(self) -> Set[LoRARequest]:
-        return {g.seq_group.lora_request for g in self.scheduled_seq_groups}
+        result = {}
```
Comment (rkooo567) [inline vllm/core/scheduler.py:143]: there was a bug where this returns {None}

File: vllm/attention/backends/torch_sdpa.py [hunk 3]
```diff
@@ -49,17 +50,14 @@ def copy_blocks(
 
 
 @dataclass
-class TorchSDPAMetadata(AttentionMetadata, PagedAttentionMetadata):
+class TorchSDPAMetadata(AttentionMetadataPerStage, PagedAttentionMetadata):
```
Comment (rkooo567) [inline vllm/attention/backends/torch_sdpa.py:53]: QQ: Currently, do we run all attention backend in tests?

File: tests/worker/test_model_runner.py [hunk 4]
```diff
@@ -190,3 +193,40 @@ def test_prepare_decode_cuda_graph(batch_size):
...
+     _, _,
+     slot_mapping) = (model_runner._prepare_prompt(seq_group_metadata_list))
+    assert len(input_tokens) == 0
+    assert len(input_positions) == 0
+    assert attn_metadata is None
+    assert len(slot_mapping) == 0
+    assert len(return_prompt_lens) == 0
+
+
+# SANG-TODO Test chunked prefill case.
```
Comment (rkooo567) [inline tests/worker/test_model_runner.py:232]: This one is WIP

**Comment by rkooo567** [comment]
Initial throughput benchmark on 7B, 1 GPU

```
# not enabled
Throughput: 1.98 requests/s, 974.93 tokens/s
# enabled
Throughput: 2.02 requests/s, 995.76 tokens/s
```

I am going to start further benchmark at better machines and higher TP next week

**Comment by scv119** [comment]
for the benchmarks, does it enable cuda-graph? also worth noting the detail of benchmarks (i.e. type of GPU and benchmark setup)

**Comment by ywang96** [comment]
@rkooo567 If this PR is ready, I'm happy to test this it with some benchmarking as well!

**Comment by rkooo567** [comment]
@ywang96 I think it may actually work. The test failures now may be just test specific issues. The basic_correctness_test seems pass when I ran it (including cuda graph + tp > 1). 

@scv119 actually the original RFC doesn't include cuda graph yet after talking with Woosuk (he mentioned he didn't see much improvement in cuda graph if batch size > 256). I will run our internal with decode-only cuda graph (oss status quo) vs whole cuda graph to see the actual impact and do a follow up. Have you observed big perf difference here?

**Comment by rkooo567** [comment]
also want to emphaisze the benchmark above runs 1 A10 GPU + llama 7B. It was just a dry run. I will do more serious benchmark next week, and we may need cuda graph for chunked prefill to (now it is enabled only when batch only contains decode)

File: vllm/worker/model_runner.py [hunk 5]
```diff
@@ -154,10 +155,9 @@ def get_max_block_per_batch(self) -> int:
     def _prepare_prompt(
         self,
         seq_group_metadata_list: List[SequenceGroupMetadata],
-    ) -> Tuple[torch.Tensor, torch.Tensor, AttentionMetadata, List[int],
-               List[int], List[int], List[int], Set[LoRARequest],
-               torch.Tensor]:
-        assert len(seq_group_metadata_list) > 0
+    ) -> Tuple[List[int], List[int], Optional[AttentionMetadataPerStage],
+               List[int], List[int], List[int], List[int], Set[LoRARequest],
+               Optional[torch.Tensor], List[int]]:
```
Comment (simon-mo) [inline vllm/worker/model_runner.py:160]: ouch... can we do named tuple or data class here. i know this not your code but this really hurts...

File: vllm/worker/model_runner.py [hunk 6]
```diff
@@ -617,6 +639,16 @@ def prepare_input_tensors(
                 lora_mapping = None
 
             # Broadcast the metadata.
+            # If batch contains both prefill and decode, it sends 2 broadcasts.
+            # If it only contains 1 type, it triggers a single broadcast.
+            if (prefill_attn_metadata is not None
+                    and decode_attn_metadata is not None):
+                batch_type = "mixed"
```
Comment (simon-mo) [inline vllm/worker/model_runner.py:646]: i think making this an IntEnum will help broadcasting to avoid pickling

File: vllm/attention/backends/torch_sdpa.py (hunk 3 above)
Comment (rkooo567) [reply vllm/attention/backends/torch_sdpa.py:53]: I will create a PR for it

File: vllm/attention/backends/xformers.py [hunk 7]
```diff
@@ -209,9 +210,28 @@ def forward(
...
+        query = query[:num_prefill_tokens]
+        key = key[:num_prefill_tokens]
+        value = value[:num_prefill_tokens]
+
+        assert query.shape[0] == num_prefill_tokens
+        assert decode_query.shape[0] == num_decode_tokens
+
+        if num_prefill_tokens > 0:
+            prefill_meta = attn_metadata.prefill_metadata
+            assert prefill_meta is not None
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply vllm/attention/backends/xformers.py:231]: moved invariant check and used warlus operator

File: vllm/config.py (hunk 1 above)
Comment (AgrawalAmey) [reply vllm/config.py:567]: actually, there is a no good default -- all depends on application requirements, model, hw, etc. I have an experimental auto-tuner which sets the maximum chunk size that can satisfy the application requirements. It is pretty straight forward, it just takes tbt target and find out the max token budget that can satisfy it.

File: vllm/attention/backends/xformers.py [hunk 8]
```diff
@@ -123,18 +115,27 @@ def __post_init__(self):
...
     Currently, prompt tokens don't contain any padding.
 
     The prompts might have different lengths, while the generation tokens
     always have length 1.
+
+    If chunked prefill is enabled, prefill tokens and decode tokens can be
+    batched together in a flattened 1D query.
+
+    |<----- num_prefill_tokens ---->|<------- num_decode_tokens ----------->|	
+    |<-prompt_0->|...|<-prompt_N-1->|<-generation_0->|...|<-generation_M-1->|
```
Comment (AgrawalAmey) [inline vllm/attention/backends/xformers.py:135]: nit: can we align the terminology?

File: vllm/attention/backends/xformers.py [hunk 9]
```diff
@@ -226,35 +243,41 @@ def forward(
...
                     key,
                     value,
                     key_cache,
                     value_cache,
-                    attn_metadata.block_tables,
-                    attn_metadata.subquery_start_loc,
-                    attn_metadata.prompt_lens_tensor,
-                    attn_metadata.context_lens,
-                    attn_metadata.max_subquery_len,
+                    prefill_meta.block_tables,
```
Comment (AgrawalAmey) [inline vllm/attention/backends/xformers.py:261]: @rkooo567 you mentioned this kernel is really slow right? should we show some warning if someone uses this backend with chunked prefills?

File: vllm/attention/backends/flash_attn.py [hunk 10]
```diff
@@ -188,52 +195,72 @@ def forward(
...
                 )
+                assert output[:num_prefill_tokens].shape == out.shape
+                output[:num_prefill_tokens] = out
             else:
                 # prefix-enabled attention
                 # TODO(Hai) this triton kernel has regression issue (broke) to
                 # deal with different data types between KV and FP8 KV cache,
                 # to be addressed separately.
-                output = PagedAttention.forward_prefix(
+                output[:num_prefill_tokens] = PagedAttention.forward_prefix(
```
Comment (AgrawalAmey) [inline vllm/attention/backends/flash_attn.py:240]: with flash_attn_varlen_func, we can do both prefill and decode attention together right?

File: vllm/attention/backends/xformers.py (hunk 9 above)
Comment (rkooo567) [reply vllm/attention/backends/xformers.py:261]: Hmm feel like that's a bit weird because there's no alternative option. I think we should probably just mark it as experimental feature for now?

Comment (rkooo567) [reply vllm/attention/backends/xformers.py:243]: I can also do some additional e2e benchmark. When I ran it with 7B, the perf was actually not very different

File: vllm/config.py [hunk 11]
```diff
@@ -563,9 +563,16 @@ def __init__(
...
+                self.max_num_batched_tokens = 768
+            else:
+                # If max_model_len is too short, use 2048 as the default value
+                # for higher throughput.
+                self.max_num_batched_tokens = max(max_model_len, 2048)
+        if enable_chunked_prefill:
+            logger.info("Chunked prefill is enabled (EXPERIMENTAL).")
```
Comment (rkooo567) [inline vllm/config.py:574]: cc @AgrawalAmey

**Comment by rkooo567** [comment]
All comments are addressed, and I believe the tests are passing now

File: vllm/attention/backends/torch_sdpa.py [hunk 12]
```diff
@@ -142,36 +140,51 @@ def forward(
...
+        decode_query = query[num_prefill_tokens:]
+        # QKV for prefill.
+        query = query[:num_prefill_tokens]
+        key = key[:num_prefill_tokens]
+        value = value[:num_prefill_tokens]
+
+        assert query.shape[0] == num_prefill_tokens
+        assert decode_query.shape[0] == num_decode_tokens
+
+        if prefill_meta := attn_metadata.prefill_metadata:
```
Comment (zhuohan123) [inline vllm/attention/backends/torch_sdpa.py:159]: Small Python style question: is `:=` widely used now? I personally prefer the more old-school way.

File: vllm/model_executor/parallel_utils/communication_op.py [hunk 13]
```diff
@@ -171,10 +171,18 @@ def broadcast_tensor_dict(
...
                 tensor = tensor_dict[key]
-                torch.distributed.broadcast(tensor, src=src, group=group)
+                async_handles.append(
+                    torch.distributed.broadcast(tensor,
+                                                src=src,
+                                                group=group,
+                                                async_op=True))
+        for async_handle in async_handles:
+            async_handle.wait()
+
```
Comment (zhuohan123) [inline vllm/model_executor/parallel_utils/communication_op.py:185]: Thanks for optimizing this!

File: vllm/worker/model_runner.py [hunk 14]
```diff
@@ -39,6 +41,39 @@
...
+
+
+class PrepareDecodeMetadata(NamedTuple):
+    input_tokens: List[int]
+    input_positions: List[int]
+    attn_metadata: Optional[AttentionMetadata]
+    lora_index_mapping: List[int]
+    lora_prompt_mapping: List[int]
+    lora_requests: Set[LoRARequest]
+    slot_mapping: List[int]
```
Comment (zhuohan123) [inline vllm/worker/model_runner.py:64]: How is chunked prefill handled here?

File: vllm/worker/model_runner.py [hunk 15]
```diff
@@ -171,6 +203,9 @@ def _prepare_prompt(
         prefix_block_tables: List[List[int]] = []
         multi_modal_input_list: List[torch.Tensor] = []
 
+        if len(seq_group_metadata_list) == 0:
+            return [], [], None, [], [], [], [], set(), None, []
```
Comment (zhuohan123) [inline vllm/worker/model_runner.py:207]: Let's wrap this with `PreparePromptMetadata`? Also why do we need a special check here?

File: vllm/worker/model_runner.py [hunk 16]
```diff
@@ -617,6 +685,16 @@ def prepare_input_tensors(
...
             # Broadcast the metadata.
+            # If batch contains both prefill and decode, it sends 2 broadcasts.
+            # If it only contains 1 type, it triggers a single broadcast.
+            if (prefill_attn_metadata is not None
+                    and decode_attn_metadata is not None):
+                batch_type = BatchType.MIXED
+            elif prefill_attn_metadata is not None:
+                batch_type = BatchType.PREFILL
+            else:
+                batch_type = BatchType.DECODE
```
Comment (zhuohan123) [inline vllm/worker/model_runner.py:696]: Why is this needed? Can we detect this in the attention kernel instead?

File: vllm/worker/model_runner.py [hunk 17]
```diff
@@ -588,26 +616,66 @@ def prepare_input_tensors(
...
+            num_decode_tokens = len(decode_input_tokens)
+
+            # Coalesce tensors. Note that attn_metadata is currently not
+            # coalesced for simplicity.
+            input_tokens.extend(decode_input_tokens)
+            input_positions.extend(decode_input_positions)
+            slot_mapping.extend(decode_slot_mapping)
+            lora_index_mapping.extend(decode_lora_index_mapping)
+            lora_prompt_mapping.extend(decode_lora_prompt_mapping)
+            lora_requests.update(decode_lora_requests)
```
Comment (zhuohan123) [inline vllm/worker/model_runner.py:667]: Again I feel the complex logic here is caused by separation of prefill and decode stage. I once thought that with chunked prefill, prompt and decode should be unified.

**Review by zhuohan123 — COMMENTED** [review]
@rkooo567 Let's chat offline about this PR? My high-level comment: with chunked prefill, we should no longer distinguish prefill and decode outside the attention layer, which should simplify the code significantly.

File: vllm/attention/backends/torch_sdpa.py (hunk 12 above)
Comment (rkooo567) [reply vllm/attention/backends/torch_sdpa.py:159]: This was based on simon's feedback. https://github.com/vllm-project/vllm/pull/3884#discussion_r1556390926. Lmk if you want to just revert this

**Comment by rkooo567** [comment]
@zhuohan123 I believe the current style is the best iterative step to merge the PR. (it is also the way it is done in our internal repo). I'd love to do unification, but I feel like we should do it step by step instead of all in the current PR;

- Some of internal metadata is not consistent between decode/prefill. For example, context_len. We should fix these tech debt.
- This requires to refactor existing code to be prefill/decode agonistic because chunked prefill co-exist with existing code at least for some time.

File: vllm/worker/model_runner.py (hunk 16 above)
Comment (rkooo567) [reply vllm/worker/model_runner.py:696]: This is to send both metadata. Other alternative is to coelesce them in a single attention metadata

File: vllm/worker/model_runner.py (hunk 15 above)
Comment (rkooo567) [reply vllm/worker/model_runner.py:207]: it is much simpler this way otherwise we should make code below to handle seq_group_metadata_list == 0 (there's additional logic outside iterating seq group metadata)

File: vllm/worker/model_runner.py (hunk 14 above)
Comment (rkooo567) [reply vllm/worker/model_runner.py:64]: This addresses https://github.com/vllm-project/vllm/pull/3884#discussion_r1556490831 comment. It is combined inside `prepare_input_tensors`

File: vllm/attention/backends/abstract.py [hunk 18]
```diff
@@ -59,6 +60,41 @@ def asdict_zerocopy(self) -> Dict[str, Any]:
...
+    num_prefill_tokens: int
+    # Number of decode tokens. Note that it is equivalent to the number of
+    # decode requests.
+    num_decode_tokens: int
+    # The attention metadata for prefill requests in a batch.
+    # None if there's no prefill requests in a batch.
+    prefill_metadata: Optional[T]
+    # The attention metadata for decode requests in a batch.
+    # None if there's no decode requests in a batch.
+    decode_metadata: Optional[T]
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply vllm/attention/backends/abstract.py:81]: My opinion is this requires quite of refactoring, which I believe it is easier to be done as a follow up. https://github.com/vllm-project/vllm/pull/3884#issuecomment-2047105929

But let's resolve it in the meeting tmrw!

**Comment by rkooo567** [comment]
Discussed offline.

- We agreed on what's "clean" state. It requires some thoughts, and I will do it as a follow up
- The sequence is logprob fix -> cuda graph enablement for prefill -> refactor this part to be clean

File: vllm/attention/backends/xformers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (zhuohan123) [reply vllm/attention/backends/xformers.py]: I'm not a fan of valrus operator in python and I believe it's [pretty controversial](https://youtu.be/KN2TTiGpDvM). I would vote for limiting it's use in vLLM. @simon-mo what do you think?

File: vllm/worker/model_runner.py (hunk 17 above)
Comment (rkooo567) [reply vllm/worker/model_runner.py:692]: discussed offline; will do refactoring to combine all metadata into a single data structure

**Review by simon-mo — APPROVED** [review]

File: vllm/attention/backends/torch_sdpa.py [hunk 19]
```diff
@@ -181,28 +194,32 @@ def forward(
...
+                decode_meta.context_lens,
+                decode_meta.max_context_len,
                 attn_metadata.kv_cache_dtype,
                 self.num_kv_heads,
                 self.scale,
                 self.alibi_slopes,
                 kv_scale,
             )
+            assert out.shape == output[num_prefill_tokens:].shape
+            output[num_prefill_tokens:]
```
Comment (jikunshang) [inline vllm/attention/backends/torch_sdpa.py:222]: should be `output[num_prefill_tokens:] = out` here?

**Comment by casper-hansen** [comment]
> It also means we can have up to maximum 2 chunked prefill at any given time.

Can we optimize this to allow many more chunks at a time? I thought the main optimization of [DeepSpeed-MII](https://github.com/microsoft/DeepSpeed-MII) was the chunked prefilling, which greatly speeds up the context processing and therefore the throughput.

File: vllm/attention/backends/torch_sdpa.py (hunk 19 above)
Comment (rkooo567) [reply vllm/attention/backends/torch_sdpa.py:222]: oops that;s correct. I don't know how the test passes...

Comment (rkooo567) [reply vllm/attention/backends/torch_sdpa.py:222]: oh it is handled https://github.com/vllm-project/vllm/commit/8afca50889bad6ad987c523c48c31fc52fcb72e4

Comment (jikunshang) [reply vllm/attention/backends/torch_sdpa.py:222]: We didn't enable much test for cpu path, we will try to add step by step later. thanks for your confirming!:)

**Comment by rkooo567** [comment]
@casper-hansen that's totally doable. Do you happen to know the perf benefits or related sections for multiple prefill chunks? I followed the policy from https://arxiv.org/abs/2308.16369. Also cc @AgrawalAmey for thoughts

**Comment by casper-hansen** [comment]
@rkooo567 Please see this blog from Microsoft.

https://github.com/microsoft/DeepSpeed/tree/master/blogs/deepspeed-fastgen#3-dynamic-splitfuse-a-novel-prompt-and-generation-composition-strategy
